reed_solomon.go

```diff
@@ -0,0 +1,154 @@
+package chunk
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"sync"
+
+	"github.com/ipfs/go-ipfs-files"
+	rs "github.com/klauspost/reedsolomon"
+)
+
+const (
+	DefaultReedSolomonDataShards   = 10
+	DefaultReedSolomonParityShards = 20
+	DefaultReedSolomonShardSize    = DefaultBlockSize
+)
+
+// reedSolomonSplitter implements the MultiSplitter interface and splits into multiple
+// Splitters based on data + parity shards. Each Splitter corresponds to one
+// Reed-Solomon shard and splits using the default SizeSplitter.
+// Reed-Solomon shard Splitters are safe for concurrent reading.
+// The default Splitter interface for ReedSolomonSplitter is a serialized
+// read of all shard chunks.
+type reedSolomonSplitter struct {
+	sync.Mutex
+
+	r         io.Reader
+	spls      []Splitter
+	splIndex  int
+	numData   uint64
+	numParity uint64
+	size      uint64
+	err       error
+}
+
+// NewReedSolomonSplitter takes in the number of data and parity chards, plus
+// a size splitting the shards and returns a ReedSolomonSplitter.
+func NewReedSolomonSplitter(r io.Reader, numData, numParity, size uint64) (
+	*reedSolomonSplitter, error) {
+	var fileSize int64
+	if fi, ok := r.(files.FileInfo); ok {
+		fileSize = fi.Stat().Size()
+	} else {
+		// Not a file object, but we need to know the full size before
+		// being streamed for reed-solomon encoding.
+		// Copy it to a buffer as a last resort.
+		b, err := ioutil.ReadAll(r)
+		if err != nil {
+			return nil, err
+		}
+		fileSize = int64(len(b))
+		// Re-pack reader
+		r = bytes.NewReader(b)
+	}
+
+	rss, err := rs.NewStreamC(int(numData), int(numParity), true, true)
+	if err != nil {
+		return nil, err
+	}
+
+	var bufs []*bytes.Buffer
+	for i := 0; i < int(numData+numParity); i++ {
+		bufs = append(bufs, &bytes.Buffer{})
+	}
+	var dataWriters []io.Writer
+	var parityWriters []io.Writer
+	for i, b := range bufs {
+		if uint64(i) < numData {
+			dataWriters = append(dataWriters, io.Writer(b))
+		} else {
+			parityWriters = append(parityWriters, io.Writer(b))
+		}
+	}
+	// Split data into even data shards first
+	err = rss.Split(r, dataWriters, fileSize)
```

Comment:
I think this method from "github.com/klauspost/reedsolomon" can be a possible bottleneck. The code inside the method is sequentially reading and writing. May need to revisit later.

Reply:
The source would have been multi-streamed in the first place - kinda hard to do. From the library's benchmark it seems pretty performant. We will limit the max file size to make sure this is not over the roof.